File: C/Sully.c
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/wait.h>
#define FT(x) int main(int argc, char** argv) { (void)argc; char fileName[] = "Sully_x.c"; char execName[] = "Sully_x"; if (argv[0][5] != '_') i++; fileName[6] = '0' + i - 1; execName[6] = '0' + i - 1; FILE* f = fopen(fileName, "w"); if (f == 0) return 1; fprintf(f, "%s%s\n#define STR(x) #x\n#define EXPAND(x) STR(x)\n\n/*\n This program will print its own source when run.\n*/\n\nint i = %i;\n\nFT(EXPAND(FT(x)))\n", "#include <fcntl.h>\n#include <stdio.h>\n#include <unistd.h>\n#include <stdlib.h>\n#include <sys/wait.h>\n#define FT(x) ", x, i - 1); fclose(f); if (!(i - 1)) return 0; int pid = fork(); if (pid == 0) { execl("/usr/bin/clang", "/usr/bin/clang", "-Werror", "-Wall", "-Wextra", fileName, "-o", execName, (void*)0); exit(0);} int status = 0; if (waitpid(pid, &status, 0) != pid || status) { return 1; } execl(execName, execName, (void*)0); return 0; }
#define STR(x) #x
#define EXPAND(x) STR(x)

/*
 This program will print its own source when run.
*/

int i = 5;

FT(EXPAND(FT(x)))

File: Python/Sully.py
import sys
import os

# This program will print its own source when run.

i = 5

codeStr = "import sys\nimport os\n\n# This program will print its own source when run.\n\ni = {1}\n\ncodeStr = \"{0}\"\n\ndef main():\n    global i\n    if sys.argv[0][5] == '_':\n        i = i -1\n    newFileName = f\"Sully_\" + str(i) + \".py\"\n    try:\n        with open(newFileName, \"w\") as file:\n            file.write(codeStr.format(codeStr.replace('\\\\', '\\\\\\\\').replace('\"', '\\\\\"').replace('\\n', '\\\\n'), i))\n    except:\n        print(\"Error writing file!\")\n        sys.exit(1)\n    if i == 0:\n        sys.exit(0)\n    os.system(f\"python3 \" + newFileName)\n\nif __name__ == \"__main__\":\n    main()\n"

def main():
    global i
    if sys.argv[0][5] == '_':
        i = i -1
    newFileName = f"Sully_" + str(i) + ".py"
    try:
        with open(newFileName, "w") as file:
            file.write(codeStr.format(codeStr.replace('\\', '\\\\').replace('"', '\\"').replace('\n', '\\n'), i))
    except:
        print("Error writing file!")
        sys.exit(1)
    if i == 0:
        sys.exit(0)
    os.system(f"python3 " + newFileName)

if __name__ == "__main__":
    main()
